pkg/logs/tail.go

```diff
@@ -0,0 +1,610 @@
+	c.CmdClause.Flag("stream", "Stream specifies which of 'stdout' or 'stderr' to output, defaults to undefined (all streams)").StringVar(&c.cfg.stream)
+
+	return &c
+}
+
+// Exec invokes the application logic for the command.
+func (c *TailCommand) Exec(in io.Reader, out io.Writer) error {
+	c.Input.Kind = fastly.ManagedLoggingInstanceOutput
+	c.cfg.path = fmt.Sprintf("%s/service/%s/log_stream/managed/instance_output", config.DefaultEndpoint, c.Input.ServiceID)
+
+	c.dieCh = make(chan struct{})
+	c.batchCh = make(chan Batch)
+	c.doneCh = make(chan struct{})
+
+	c.hClient = http.DefaultClient
+	c.token, _ = c.Globals.Token()
+
+	// Adjust the from/to times if they are
+	// defined. We adjust the times based on searchPadding.
+	c.adjustTimes()
+
+	// Enable managed logging if not already enabled.
+	if err := c.enableManagedLogging(out); err != nil {
+		return err
+	}
+
+	sigs := make(chan os.Signal, 2)
+	signal.Notify(sigs, os.Interrupt, syscall.SIGTERM, syscall.SIGHUP)
+
+	// Start the output loop.
+	go c.outputLoop(out)
+
+	// Start tailing the logs.
+	go c.tail(out)
+
+	<-sigs
+	close(c.dieCh)
+
+	return nil
+}
+
+//
+// Client
+//
+
+// Tail starts the virtual tail process. Tail fetches data from the eventbuffer
+// API. It hands off the requested logs to the outputloop for the actual
+// printing.
+func (c *TailCommand) tail(out io.Writer) {
+	// Start this with --from and --to if set.
+	curWindow := c.cfg.from
+	toWindow := c.cfg.to
+
+	// Start the loop with an initial address to query.
+	path, err := makeNewPath(c.cfg.path, curWindow, "")
+	if err != nil {
+		text.Error(out, "error starting tail: %v", err)
+	}
+
+	// lastBatchID keeps the last successfully read Batch.ID in case we need
+	// re-request on failure.
+	var lastBatchID string
+
+	for {
+		// Check to see if we already passed the "to" requirement.
+		if toWindow != 0 && curWindow > toWindow {
+			text.Info(out, "Reached window: %v which is newer than the requested 'to': %v", curWindow, toWindow)
+			// We are done, but we still want printing to finish.
+			close(c.doneCh)
+			break
+		}
+
+		req, err := http.NewRequest("GET", path, nil)
+		req.Header.Add("Fastly-Key", c.token)
+		if err != nil {
+			text.Error(out, "unable to create new request: %v", err)
+			os.Exit(1)
+		}
+
+		resp, err := c.doReq(req)
+		if err != nil {
+			text.Error(out, "unable to request: %v", err)
+			os.Exit(1)
+		}
+
+		// Check that our request was successful. If the server is
+		// having trouble, retry after waiting for some time.
+		if resp.StatusCode != http.StatusOK {
+			// In an effort to clean up the output, do not print on
+			// 503's.
+			if resp.StatusCode != http.StatusServiceUnavailable {
+				text.Warning(out, "non-200 resp %d", resp.StatusCode)
+			}
+			io.Copy(ioutil.Discard, resp.Body)
+			// Try the response again after a 1 second wait.
+			if resp.StatusCode/100 == 5 && resp.StatusCode != 501 ||
```

Comment:
Do we want to implement a form of exponential backoff? (not necessary for this PR but just a thought)

Reply:
Possibly, but the window size is only 10 seconds, so it might be overkill atm.